File: iosApp/ExchangeConvertApp/QuickExchangeSection.swift
import SwiftUI
import SharedKMP

/// QuickExchangeSection component for currency conversion
/// Uses API rates for calculation: result = amount * (rates[toCurrency] / rates[fromCurrency])
struct QuickExchangeSection: View {
    let baseCurrency: String
    let rates: [String: Double]
    let onConvertClick: (Double, String, String) -> Void
    let onFromCurrencyClick: () -> Void
    let onToCurrencyClick: () -> Void
    
    @State private var amount: String = ""
    @State private var fromCurrency: String
    @State private var toCurrency: String = "EUR"
    @State private var result: String = "0.00"
    
    init(
        baseCurrency: String = "USD",
        rates: [String: Double] = [:],
        onConvertClick: @escaping (Double, String, String) -> Void = { _, _, _ in },
        onFromCurrencyClick: @escaping () -> Void = {},
        onToCurrencyClick: @escaping () -> Void = {}
    ) {
        self.baseCurrency = baseCurrency
        self.rates = rates
        self.onConvertClick = onConvertClick
        self.onFromCurrencyClick = onFromCurrencyClick
        self.onToCurrencyClick = onToCurrencyClick
        _fromCurrency = State(initialValue: baseCurrency)
    }
    
    var body: some View {
        VStack(spacing: 20) {
            // Header
            VStack(alignment: .leading, spacing: 8) {
                Text("Quick Exchange")
                    .font(.system(size: 20, weight: .bold))
                    .foregroundColor(DesignColors.onBackground)
                
                Text("Convert currencies using live rates")
                    .font(.system(size: 12, weight: .regular))
                    .foregroundColor(DesignColors.onSurfaceMuted)
            }
            
            // Amount Input
            VStack(alignment: .leading, spacing: 8) {
                Text("Amount")
                    .font(.system(size: 14, weight: .medium))
                    .foregroundColor(DesignColors.onSurfaceMuted)
                
                TextField("0.00", text: $amount)
                    .keyboardType(.decimalPad)
                    .font(.system(size: 20, weight: .semibold))
                    .foregroundColor(DesignColors.onBackground)
                    .padding(16)
                    .background(
                        RoundedRectangle(cornerRadius: 12)
                            .fill(DesignColors.surfaceVariant)
                            .overlay(
                                RoundedRectangle(cornerRadius: 12)
                                    .stroke(DesignColors.stroke.opacity(0.15), lineWidth: 1)
                            )
                    )
            }
            
            // Currency Row (From/To)
            HStack(spacing: 12) {
                // From Currency
                CurrencyButton(
                    currency: fromCurrency,
                    onClick: onFromCurrencyClick
                )
                
                // To Currency
                CurrencyButton(
                    currency: toCurrency,
                    onClick: onToCurrencyClick
                )
            }
            
            // Result Display
            HStack {
                Text("Converted Amount")
                    .font(.system(size: 12, weight: .regular))
                    .foregroundColor(DesignColors.onSurfaceMuted)
                
                Spacer()
                
                Text(result)
                    .font(.system(size: 28, weight: .bold))
                    .foregroundColor(DesignColors.onBackground)
            }
            .padding(16)
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(DesignColors.buttonBackground.opacity(0.2))
            )
            
            // Convert Button
            Button(action: {
                calculateConversion()
                if let amountValue = Double(amount) {
                    onConvertClick(amountValue, fromCurrency, toCurrency)
                }
            }) {
                Text("Convert")
                    .font(.system(size: 16, weight: .bold))
                    .foregroundColor(DesignColors.success)
                    .frame(maxWidth: .infinity)
                    .frame(height: 56)
            }
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(DesignColors.success.opacity(0.2))
                    .overlay(
                        RoundedRectangle(cornerRadius: 12)
                            .fill(DesignColors.success)
                    )
            )
        }
        .padding(24)
        .background(DesignColors.cardBackground)
        .cornerRadius(16)
    }
    
    private func calculateConversion() {
        guard let amountValue = Double(amount), amountValue > 0 else {
            result = "0.00"
            return
        }
        
        if !rates.isEmpty {
            // Formula: amount * (rates[toCurrency] / rates[fromCurrency])
            // If fromCurrency is base, use rate directly
            let fromRate = fromCurrency == baseCurrency ? 1.0 : (rates[fromCurrency] ?? 1.0)
            let toRate = toCurrency == baseCurrency ? 1.0 : (rates[toCurrency] ?? 1.0)
            let convertedAmount = amountValue * (toRate / fromRate)
            result = String(format: "%.2f", convertedAmount)
        } else {
            result = "0.00"
        }
    }
}

/// Currency button component for From/To selection
struct CurrencyButton: View {
    let currency: String
    let onClick: () -> Void
    
    var body: some View {
        Button(action: onClick) {
            HStack {
                Text(currency)
                    .font(.system(size: 16, weight: .semibold))
                    .foregroundColor(DesignColors.onBackground)
                
                Spacer()
                
                Image(systemName: "chevron.down")
                    .font(.system(size: 10))
                    .foregroundColor(DesignColors.onSurfaceMuted)
            }
            .padding(.horizontal, 16)
            .frame(height: 56)
            .background(
                RoundedRectangle(cornerRadius: 12)
                    .fill(DesignColors.surfaceVariant)
                    .overlay(
                        RoundedRectangle(cornerRadius: 12)
                            .stroke(DesignColors.stroke.opacity(0.15), lineWidth: 1)
                    )
            )
        }
    }
}

#Preview {
    QuickExchangeSection(baseCurrency: "USD", rates: ["EUR": 0.92, "GBP": 0.79])
        .padding()
        .background(DesignColors.background)
}

File: app/src/main/java/dali/hamza/echangecurrencyapp/ui/compose/component/QuickExchangeSection.kt
package dali.hamza.echangecurrencyapp.ui.compose.component

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_button_background
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_card_background
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_on_background
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_on_surface_muted
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_stroke
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_success
import dali.hamza.echangecurrencyapp.ui.compose.theme.design_surface_variant

/**
 * QuickExchangeSection component for currency conversion
 * Uses API rates for calculation: result = amount * (rates[toCurrency] / rates[fromCurrency])
 *
 * @param baseCurrency The current base currency from API
 * @param rates Map of currency codes to their exchange rates from API
 * @param onConvertClick Callback when convert button is clicked
 * @param onFromCurrencyClick Callback to open currency picker for "from" currency
 * @param onToCurrencyClick Callback to open currency picker for "to" currency
 */
@Composable
fun QuickExchangeSection(
    baseCurrency: String = "USD",
    rates: Map<String, Double> = emptyMap(),
    onConvertClick: (Double, String, String) -> Unit = { _, _, _ -> },
    onFromCurrencyClick: () -> Unit = {},
    onToCurrencyClick: () -> Unit = {},
    modifier: Modifier = Modifier
) {
    var amount by remember { mutableStateOf("") }
    var fromCurrency by remember { mutableStateOf(baseCurrency) }
    var toCurrency by remember { mutableStateOf("EUR") }
    var result by remember { mutableStateOf<String?>(null) }
    
    // Calculate conversion when amount or currencies change
    fun calculateConversion() {
        val amountValue = amount.toDoubleOrNull() ?: 0.0
        if (amountValue > 0 && rates.isNotEmpty()) {
            // Formula: amount * (rates[toCurrency] / rates[fromCurrency])
            // If fromCurrency is base, use rate directly
            val fromRate = if (fromCurrency == baseCurrency) 1.0 else rates[fromCurrency] ?: 1.0
            val toRate = if (toCurrency == baseCurrency) 1.0 else rates[toCurrency] ?: 1.0
            val convertedAmount = amountValue * (toRate / fromRate)
            result = String.format("%.2f", convertedAmount)
        } else {
            result = "0.00"
        }
    }

    Card(
        modifier = modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp),
        shape = RoundedCornerShape(16.dp),
        colors = CardDefaults.cardColors(
            containerColor = design_card_background
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = 0.dp)
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(design_card_background)
                .padding(24.dp),
            verticalArrangement = Arrangement.spacedBy(20.dp)
        ) {
            // Header
            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = "Quick Exchange",
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold,
                    color = design_on_background
                )
                Text(
                    text = "Convert currencies using live rates",
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Normal,
                    color = design_on_surface_muted
                )
            }

            // Amount Input
            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text(
                    text = "Amount",
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                    color = design_on_surface_muted
                )
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(56.dp)
                        .clip(RoundedCornerShape(12.dp))
                        .background(design_surface_variant)
                        .background(
                            Color(0xFF44474d).copy(alpha = 0.15f),
                            RoundedCornerShape(12.dp)
                        )
                        .padding(horizontal = 16.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = amount.ifEmpty { "0.00" },
                        fontSize = 20.sp,
                        fontWeight = FontWeight.SemiBold,
                        color = design_on_background
                    )
                }
            }

            // Currency Row (From/To)
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // From Currency
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    ExchangeButton(
                        currency = fromCurrency,
                        onClick = onFromCurrencyClick
                    )
                }

                // To Currency
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    ExchangeButton(
                        currency = toCurrency,
                        onClick = onToCurrencyClick
                    )
                }
            }

            // Result Display
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(12.dp))
                    .background(design_button_background.copy(alpha = 0.2f))
                    .padding(16.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = "Converted Amount",
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Normal,
                    color = design_on_surface_muted
                )
                Text(
                    text = result ?: "0.00",
                    fontSize = 28.sp,
                    fontWeight = FontWeight.Bold,
                    color = design_on_background
                )
            }

            // Convert Button
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(56.dp)
                    .clip(RoundedCornerShape(12.dp))
                    .background(design_success.copy(alpha = 0.2f))
                    .clickable {
                        calculateConversion()
                        val amountValue = amount.toDoubleOrNull() ?: 0.0
                        onConvertClick(amountValue, fromCurrency, toCurrency)
                    }
                    .background(design_success, RoundedCornerShape(12.dp)),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = "Convert",
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    color = design_success
                )
            }
        }
    }
}

@Composable
private fun ExchangeButton(
    currency: String,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(56.dp)
            .clip(RoundedCornerShape(12.dp))
            .background(design_surface_variant)
            .background(
                Color(0xFF44474d).copy(alpha = 0.15f),
                RoundedCornerShape(12.dp)
            )
            .clickable { onClick() }
            .padding(horizontal = 16.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = currency,
            fontSize = 16.sp,
            fontWeight = FontWeight.SemiBold,
            color = design_on_background
        )
        // Dropdown arrow could be added here
        Text(
            text = "▼",
            fontSize = 10.sp,
            color = design_on_surface_muted
        )
    }
}
